Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and type names that contain `public
+static void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll!StartupHookNamespace.StartupHookType1;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
```

Comment:
In its current form, I don't think relative paths are useful for startup hooks because the current directory is typically not under the control of the entity doing the hooking (and will be arbitrary).   

On the other hand, having a hook that is relative to app directory is useful (if the user puts the file in that directory they can give a relative path to it).    Thus we may wish to define the current directory as the app directory (Basically you change to this directory before resolving paths, then change back). 

I am trying to decide if this scenario is actually worthwhile...

Reply:
>  having a hook that is relative to app directory is useful 

I would agree that this is useful for the config file case. Hard to see the usefulness for the env variable case.

Reply:
> the current directory is typically not under the control of the entity doing the hooking

The working directory at process launch is controlled by whoever is launching the app, which is the scenario I had in mind.

> if the user puts the file in that directory they can give a relative path to it

I'm hoping that the user will tend not put the startup hook in the app directory. The case I envision is having a "global" startup hook inject code into different .NET processes on launch.

The case where I can see a relative path being useful is the following:
```
DOTNET_STARTUP_HOOKS='./startupHook/hook.dll' ./app/bin/main.dll
```
It could make it easier to launch the process from the command line if the hook dll is in a location close to the current working directory. However, given that this is supposed to be a low-level hook, not one frequently used by non-devs, I think it's fine to require absolute paths in this scenario.